[Driver] SR-2660: Enable the driver to accept multiple swiftmodules as linker inputs

matthewcarroll opened the pull request:
3624c99b76f5f317217be361136bfef0c1eb9dc9 implements passing swiftmodule inputs to the linker unless the -emit-module option is set.

22a39080e29c8ad902bfc6b370e829ba14167a53  decouples generating a module from setting the -g option.

New tests are added in `multiple_swiftmodule_inputs.swift`.

As mentioned in SR-2660, several tests checking for a generated module when the -g option is set had to be removed since -g is decoupled now from module generation.

Generating a non-top level module is still required to preserve implicit module wrapping of modules whose objects are ELF.

Let me know if you'd like for more tests to be added.

https://bugs.swift.org/browse/SR-2660

jrose-apple commented on lib/Driver/Driver.cpp:
```diff
@@ -1126,10 +1126,10 @@ void Driver::buildOutputInfo(const ToolChain &TC, const DerivedArgList &Args,
+  } else if (Args.hasArg(options::OPT_emit_objc_header,
+                         options::OPT_emit_objc_header_path) ||
+             (OI.shouldLink() &&
+              TC.getTriple().getObjectFormat() == llvm::Triple::ELF)) {
```
> Generating a non-top level module is still required to preserve implicit module wrapping of modules whose objects are ELF.

I'm pretty sure this feature is still only for debugging; it's just how we do that on Linux where dSYMs aren't the usual mechanism. So you should be able to drop this.

jrose-apple commented on lib/Driver/Driver.cpp:
```diff
@@ -1516,6 +1510,12 @@ void Driver::buildActions(const ToolChain &TC,
       } else
         Actions.push_back(MergeModuleAction.release());
     }
+    else {
```
Style nitpick: please put `else` on the same line as the close-brace.

jrose-apple commented on lib/Driver/ToolChains.cpp:
```diff
@@ -1045,19 +1045,17 @@ toolchains::Darwin::constructInvocation(const LinkJobAction &job,
+  addInputsOfType(Arguments, context.InputActions, types::TY_SwiftModuleFile);
+  
+  if (Arguments.size() > argCount) {
+    Arguments.insert(Arguments.end() - 1, "-add_ast_path");
```
This has to be present before *every* swiftmodule input, so either there needs to be a more complicated transform here or the assertion needs to stay.

jrose-apple commented on test/Driver/actions.swift:
```diff
@@ -39,9 +39,8 @@
-// DEBUG: 3: link, {1, 2}, image
-// DEBUG: 4: generate-dSYM, {3}, dSYM
+// DEBUG: 2: link, {1}, image
+// DEBUG: 3: generate-dSYM, {2}, dSYM
```
Hm. Looking at this I think `-g` + linking + `.swift` inputs should still result in a swiftmodule being generated. Otherwise, you have to specify `-emit-module` as a primary output even if you don't need one.

jrose-apple commented on test/Driver/multiple_swiftmodule_inputs.swift:
```diff
@@ -0,0 +1,100 @@
+// RUN: %swiftc_driver -c -emit-module-path single_int.swiftmodule -module-name single_int -parse-as-library %S/Inputs/single_int.swift
```
Please don't actually run these tests; just use `-driver-print-jobs` (alias `-###`) to see what they *would* do. That lets us run the tests even in cross-compilation contexts.

jrose-apple commented on test/Driver/multiple_swiftmodule_inputs.swift:
```diff
@@ -0,0 +1,100 @@
+// EMIT-MODULE: -o single_int.swiftmodule
+
+
+// RUN: %swiftc_driver -### multiple_swiftmodule_inputs.o single_int.swiftmodule 2>&1 | %FileCheck %s -check-prefix=LINK-OBJECT-AND-MODULE
```
Please don't rely on the host platform being Darwin; use a target triple here too. `-target x86_64-apple-macosx10.10` is a pretty reasonable one.

jrose-apple commented on test/Driver/unknown-inputs.swift:
```diff
@@ -11,7 +11,7 @@
 
 // RUN: %swiftc_driver -driver-print-actions %t/empty 2>&1 | %FileCheck -check-prefix=LINK-%target-object-format %s
-// RUN: not %swiftc_driver -driver-print-actions %t/empty.swiftmodule 2>&1 | %FileCheck -check-prefix=ERROR %s
+// RUN: %swiftc_driver -driver-print-actions %t/empty.swiftmodule
```
Hm. Maybe we should warn in this case? An archive containing/referencing a single swiftmodule seems like it'd usually be a mistake.

jrose-apple reviewed (commented):
Thanks for working on this, Matthew! Tagging in @dmishe, who's interested in the feature on Linux, @modocache, who took a stab at it earlier, and @aciidb0mb3r, who I believe is working around not having this in the package manager.

The main concern I have here is that the Linux toolchain hasn't been updated at all. Does everything just work? Or do we silently only get one of the modules embedded in the linked binary?

matthewcarroll commented:
I'll make those changes, and add and verify tests on Linux.

Thanks!

dmishe commented:
This is a welcome change, but a tiny correction - I don't need this on Linux, just macOS. That's probably not changing anything :)

matthewcarroll committed:
[Driver] SR-2660: Enable the driver to accept multiple swiftmodules as linker inputs

This implements passing swiftmodule inputs to the linker unless the -emit-module option is set. Swiftmodule inputs are accepted with or without the -g option. A warning is issued if all inputs are swfitmodules.

https://bugs.swift.org/browse/SR-2660

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
Did you get the chance to test this on Linux?

matthewcarroll commented:
Yes, all tests passed for me. The failures above were due to a clock sync error in Jenkins. The tests failed almost immediately, then reran and succeeded.

jrose-apple commented:
Ah, "all the tests passed" wasn't the important testing. There's still no test that says we'll actually get all of the modules embedded in the binary on Linux.

jrose-apple commented:
cc @modocache again, who's been doing some of this recently and who I haven't insisted on an equivalent test for because he hasn't said he's done yet!

modocache commented:
Oh wow, sorry I missed that there was already a pull request out for this one. Yes, I'll try to add an execution test.

CodaFi commented:
I think this change is still good to take with an updated pull request.

What do you think @jrose-apple?

CodaFi commented:
The LLDB side of things appears to be have been resolved.  The Swift Driver side is still up in the air.  I'm going to close this due to age.  If you'd like to pick it back up again, please send us an updated pull request that responds to Jordan's feedback.

Thanks!
